src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt

```diff
@@ -90,6 +90,24 @@ class RsErrorAnnotator : Annotator, HighlightRangeExtension {
         checkReferenceIsPublic(field, o, holder)
     }
 
+    private fun checkConstantUsageIsSafe(ref: RsReferenceElement, holder: AnnotationHolder) {
+        val element = ref.reference.resolve() as? RsConstant ?: return
+        val expr = ref.ancestorOrSelf<RsExpr>() ?: return
+
+        val isInUnsafeCx = expr.isInUnsafeBlockOrFn()
+        if (isInUnsafeCx) {
+            return
+        }
+
+        if (element.kind == RsConstantKind.MUT_STATIC) {
+            RsDiagnostic.UnsafeError(expr, "Use of mutable static is unsafe and requires unsafe function or block")
+                .addToHolder(holder)
+        } else if (element.kind == RsConstantKind.STATIC && element.parent is RsForeignModItem) {
+            RsDiagnostic.UnsafeError(expr, "Use of extern static is unsafe and requires unsafe function or block")
+                .addToHolder(holder)
+        }
```

Comment:
Both blocks look very similar - only one difference in error message
What do you think about moving out all same code?
Something like this:
```kotlin
val constantType = when {
    element.kind == RsConstantKind.MUT_STATIC -> "mutable"
    element.kind == RsConstantKind.STATIC && element.parent is RsForeignModItem -> "extern"
    else -> return
}
RsDiagnostic.UnsafeError(expr, "Use of $constantType static is unsafe and requires unsafe function or block")
    .addToHolder(holder)
```

Reply:
That's a good idea.  I've addressed this plus your other comment (and moved the `constantType` check above the unsafe block, to avoid the PSI traversal if the constant type isn't relevant) and rebased the tree into a few separate logical changes.